src/repr/src/row.rs

```diff
@@ -402,10 +415,142 @@ mod columnation {
     }
 }
 
+mod columnar {
+    use crate::{Row, RowRef};
+    use columnar::{
+        AsBytes, Clear, Columnar, Container, FromBytes, HeapSize, Index, IndexAs, Len, Push,
+    };
+    use mz_ore::cast::CastFrom;
+
+    #[derive(Copy, Clone, Debug, Default, PartialEq, serde::Serialize, serde::Deserialize)]
+    pub struct Rows<BC = Vec<u64>, VC = Vec<u8>> {
+        /// Bounds container; provides indexed access to offsets.
+        pub bounds: BC,
+        /// Values container; provides slice access to bytes.
+        pub values: VC,
+    }
+
+    impl Columnar for Row {
+        type Ref<'a> = &'a RowRef;
+        fn copy_from(&mut self, other: Self::Ref<'_>) {
+            self.clear();
+            self.data.extend_from_slice(other.data());
+        }
+        fn into_owned(other: Self::Ref<'_>) -> Self {
+            other.to_owned()
+        }
+        type Container = Rows;
+    }
+
+    impl<'b, BC: Container<u64>> Container<Row> for Rows<BC, &'b [u8]> {
+        type Borrowed<'a>
+            = Rows<BC::Borrowed<'a>, &'a [u8]>
+        where
+            Self: 'a;
+        fn borrow<'a>(&'a self) -> Self::Borrowed<'a> {
+            Rows {
+                bounds: self.bounds.borrow(),
+                values: self.values,
+            }
+        }
+    }
+    impl<BC: Container<u64>> Container<Row> for Rows<BC, Vec<u8>> {
+        type Borrowed<'a>
+            = Rows<BC::Borrowed<'a>, &'a [u8]>
+        where
+            BC: 'a;
+        fn borrow<'a>(&'a self) -> Self::Borrowed<'a> {
+            Rows {
+                bounds: self.bounds.borrow(),
+                values: self.values.borrow(),
+            }
+        }
+    }
+
+    impl<'a, BC: AsBytes<'a>, VC: AsBytes<'a>> AsBytes<'a> for Rows<BC, VC> {
+        fn as_bytes(&self) -> impl Iterator<Item = (u64, &'a [u8])> {
+            self.bounds.as_bytes().chain(self.values.as_bytes())
+        }
+    }
+    impl<'a, BC: FromBytes<'a>, VC: FromBytes<'a>> FromBytes<'a> for Rows<BC, VC> {
+        fn from_bytes(bytes: &mut impl Iterator<Item = &'a [u8]>) -> Self {
+            Self {
+                bounds: FromBytes::from_bytes(bytes),
+                values: FromBytes::from_bytes(bytes),
+            }
+        }
+    }
+
+    impl<BC: Len, VC> Len for Rows<BC, VC> {
+        #[inline(always)]
+        fn len(&self) -> usize {
+            self.bounds.len()
+        }
+    }
+
+    impl<'a, BC: Len + IndexAs<u64>> Index for Rows<BC, &'a [u8]> {
+        type Ref = &'a RowRef;
+        #[inline(always)]
+        fn get(&self, index: usize) -> Self::Ref {
+            let lower = if index == 0 {
+                0
+            } else {
+                self.bounds.index_as(index - 1)
+            };
+            let upper = self.bounds.index_as(index);
+            let lower: usize = lower.try_into().unwrap();
+            let upper: usize = upper.try_into().unwrap();
+            RowRef::from_slice(&self.values[lower..upper])
```

Comment:
For my own understanding: This range slicing is the reason why we impl the traits for both `&[u8]` and `Vec<u8>`, instead of being generic over `VC: Container<u8>`?

Reply:
Yep, it doesn't seem possible to capture this as a `VC: std::ops::Index` constraint because one is a reference, the other an owned thing.